File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/Module.java
package net.ccbluex.liquidbounce.features.module;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.event.Listenable;
import net.ccbluex.liquidbounce.ui.client.hud.element.elements.notifications.Notification;
import net.ccbluex.liquidbounce.utils.MinecraftInstance;
import net.ccbluex.liquidbounce.utils.render.ChatColor;
import net.ccbluex.liquidbounce.value.Value;
import net.minecraft.client.audio.PositionedSoundRecord;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

/**
 * LiquidBounce Hacked Client
 * A minecraft forge injection client using Mixin
 *
 * @game Minecraft
 * @author CCBlueX
 */
@SideOnly(Side.CLIENT)
public class Module extends MinecraftInstance implements Listenable {

    protected String name = getClass().getAnnotation(ModuleInfo.class).name();
    protected String description = getClass().getAnnotation(ModuleInfo.class).description();
    protected ModuleCategory category = getClass().getAnnotation(ModuleInfo.class).category();
    protected boolean state;
    private int keyBind = getClass().getAnnotation(ModuleInfo.class).keyBind();
    private final boolean canEnable = getClass().getAnnotation(ModuleInfo.class).canEnable();

    // HUD
    public final float hue = (float) Math.random();
    public float slide;

    public String getName() {
        return name;
    }

    public String getTagName() {
        return getName() + (getTag() == null ? "" : " §7" + getTag());
    }

    public String getColorlessTagName() {
        return getName() + (getTag() == null ? "" : " " + ChatColor.stripColor(getTag()));
    }

    public String getDescription() {
        return description;
    }

    public ModuleCategory getCategory() {
        return category;
    }

    public boolean getState() {
        return state;
    }

    public int getKeyBind() {
        return keyBind;
    }

    public void setKeyBind(final int keyBind) {
        this.keyBind = keyBind;
        LiquidBounce.CLIENT.fileManager.saveConfig(LiquidBounce.CLIENT.fileManager.modulesConfig);
    }

    public String getTag() {
        return null;
    }

    public void setState(final boolean state) {
        try {
            if(getState() == state)
                return;

            onToggle(state);

            if(state) {
                onEnable();
                if(canEnable)
                    this.state = true;
            }else{
                onDisable();
                this.state = false;
            }

            LiquidBounce.CLIENT.fileManager.saveConfig(LiquidBounce.CLIENT.fileManager.modulesConfig);
        }catch(final Exception e) {
            e.printStackTrace();
        }
    }

    public void toggle() {
        setState(!getState());
    }

    public void onToggle(final boolean state) {
        if(!LiquidBounce.CLIENT.isStarting && getState() != state) {
            mc.getSoundHandler().playSound(PositionedSoundRecord.create(new ResourceLocation("random.click"), 1.0F));
            LiquidBounce.CLIENT.hud.addNotification(new Notification((state ? "Enabled " : "Disabled ") + getName()));
        }
    }

    public void onEnable() {
    }

    public void onDisable() {
    }

    public void onStarted() {
    }

    public boolean showArray() {
        return true;
    }

    // Value

    public Value getValue(final String valueName) {
        for(final Field field : getClass().getDeclaredFields()) {
            try {
                field.setAccessible(true);

                final Object o = field.get(this);

                if(o instanceof Value) {
                    final Value value = (Value) o;

                    if(value.getName().equalsIgnoreCase(valueName))
                        return value;
                }
            }catch(IllegalAccessException e) {
                e.printStackTrace();
            }
        }

        return null;
    }

    public List<Value> getValues() {
        final List<Value> values = new ArrayList<>();

        for(final Field field : getClass().getDeclaredFields()) {
            try {
                field.setAccessible(true);

                final Object o = field.get(this);

                if(o instanceof Value)
                    values.add((Value) o);
            }catch(IllegalAccessException e) {
                e.printStackTrace();
            }
        }

        return values;
    }

    @Override
    public boolean handleEvents() {
        return getState();
    }
}

File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/Module.kt
package net.ccbluex.liquidbounce.features.module

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.event.Listenable
import net.ccbluex.liquidbounce.ui.client.hud.element.elements.notifications.Notification
import net.ccbluex.liquidbounce.utils.MinecraftInstance
import net.ccbluex.liquidbounce.utils.render.ColorUtils.stripColor
import net.ccbluex.liquidbounce.value.Value
import net.minecraft.client.audio.PositionedSoundRecord
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.util.*

/**
 * LiquidBounce Hacked Client
 * A minecraft forge injection client using Mixin
 *
 * @game Minecraft
 * @author CCBlueX
 */
@SideOnly(Side.CLIENT)
open class Module : MinecraftInstance(), Listenable {
    private val canEnable = javaClass.getAnnotation(ModuleInfo::class.java).canEnable
    var name = javaClass.getAnnotation(ModuleInfo::class.java).name
    var description = javaClass.getAnnotation(ModuleInfo::class.java).description
    var category = javaClass.getAnnotation(ModuleInfo::class.java).category
    var keyBind = javaClass.getAnnotation(ModuleInfo::class.java).keyBind
        set(keyBind) {
            field = keyBind
            LiquidBounce.CLIENT.fileManager.saveConfig(LiquidBounce.CLIENT.fileManager.modulesConfig)
        }
    var state = false
        set(value) {
            if (field == value) {
                return
            }
            onToggle(value)
            if (value) {
                onEnable()
                if (canEnable)
                    field = true
            } else {
                onDisable()
                field = false
            }
            LiquidBounce.CLIENT.fileManager.saveConfig(LiquidBounce.CLIENT.fileManager.modulesConfig)
        }


    // HUD
    val hue = Math.random().toFloat()
    var slide = 0f

    val tagName: String
        get() = name + if (tag == null) "" else " §7$tag"

    val colorlessTagName: String
        get() = name + if (tag == null) "" else " " + stripColor(tag)

    open val tag: String?
        get() = null

    fun toggle() {
        state = !state
    }

    open fun onToggle(state: Boolean) {
        if (!LiquidBounce.CLIENT.isStarting && this.state != state) {
            mc.soundHandler.playSound(PositionedSoundRecord.create(ResourceLocation("random.click"), 1.0f))
            LiquidBounce.CLIENT.hud.addNotification(Notification((if (state) "Enabled " else "Disabled ") + name))
        }
    }

    open fun onEnable() {}
    open fun onDisable() {}

    fun onStarted() {}
    open fun showArray(): Boolean {
        return true
    }

    // Value
    open fun getValue(valueName: String?): Value<*>? {
        for (field in javaClass.declaredFields) {
            try {
                field.isAccessible = true
                val o = field[this]
                if (o is Value<*>) {
                    if (o.name.equals(valueName, ignoreCase = true)) {
                        return o
                    }
                }
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }
        return null
    }

    open val values: MutableList<Value<*>>
        get() {
            val values: MutableList<Value<*>> = ArrayList()
            for (_field in javaClass.declaredFields) {
                try {
                    _field.isAccessible = true
                    val o = _field[this]
                    if (o is Value<*>) {
                        values.add(o)
                    }
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            }
            return values
        }

    override fun handleEvents(): Boolean {
        return state
    }
}